parsil/src/executor.rs

```diff
@@ -680,6 +680,120 @@ impl<'a, C: ContextProvider> AstMutator for Executor<'a, C> {
+        let mut key_fetcher = KeyFetcher {
+            settings: &mut self.settings,
+        };
+        key_fetcher.post_table_factor(table_factor)
+    }
+
+    fn post_select(&mut self, select: &mut Select) -> Result<()> {
+        let replace_wildcard = || {
+            // we expand the Wildcard by replacing it will all the columns of the original table
+            assert_eq!(select.from.len(), 1); // single table queries
+            let table = &select.from.first().unwrap().relation;
+            match table {
+                TableFactor::Derived {
+                    lateral,
+                    subquery,
+                    alias,
+                } => {
+                    subquery
+                        .as_ref()
+                        .body
+                        .as_ref()
+                        .as_select()
+                        .unwrap()
+                        .projection
+                        .iter()
+                        .filter_map(|item| {
+                            let expr = match item {
+                                SelectItem::ExprWithAlias { expr, alias } => {
+                                    Expr::Identifier(alias.clone())
+                                }
+                                SelectItem::UnnamedExpr(expr) => expr.clone(),
+                                _ => unreachable!(),
+                            };
+                            // we need to filter out KEY and EPOCH from the columns expanded by the Wildcard,
+                            // as these ones are the columns over which we need to apply DISTINCT
+                            match &expr {
+                                Expr::Identifier(ident)
+                                    if ident.value == EPOCH || ident.value == KEY =>
+                                {
+                                    None
+                                }
+                                _ => Some(expr),
+                            }
+                        })
+                        .collect::<Vec<_>>()
+                }
+                _ => unreachable!(), // post_table_factor makes `TableFactor::Derived`
+            }
+        };
+        // need to:
+        // 1. add KEY and EPOCH to existing `SelectItem`s
+        // 2. Ensure that, if there is DISTINCT keyword in the original query,
+        //    the original `SelectItem`s are wrapped in `DISTINCT ON`, to
+        //    ensure that we return only DISTINCT results
+        // first, turn existing `SelectItem`s in a vector of Expressions
+        if let Some(distinct) = select.distinct.as_mut() {
+            let items = select
+                .projection
+                .iter()
+                .flat_map(|item| {
+                    match item {
+                        SelectItem::UnnamedExpr(expr) => vec![expr.clone()],
+                        SelectItem::ExprWithAlias { expr, alias } => vec![expr.clone()], // we don't care about alias here
```

Comment:
Is there a reason to mask the aliasing here; could it be possible to just let the `SelectItem` be “as is”?

Reply:
Because the `DISTINCT ON` takes as input a vector of expressions, not `SelectItem`s. Don't know if there is another way to take into account the alias. Wdyt?

Reply:
Nevermind, I misunderstood the objective. All good for me then.